### DISCUSSION TITLE
QSPI / MIPI Support for LovyanGFX

### DISCUSSION THREAD

**Discussion by tobozo** [lovyan03/LovyanGFX#663] [Device Configurations]
This post was created in order to keep track of the progress with all QSPI panel device requests and related information.

You can contribute to the discussion by extending the devices/products list.

Valid QSPI support for LovyanGFX only exists as a proof of concept on a [heavily modified fork](https://github.com/Forairaaaaa/Monica/tree/main/Firmware/SourceCode/components/LovyanGFX/src/lgfx/v1/platforms/esp32).

An attempt was made at merging those changes back into LovyanGFX (see [QSPI branch](https://github.com/lovyan03/LovyanGFX/tree/QSPI)), but the arduino SPI library and the esp-idf underlying SPI functions evolved since the Monica project was created, and the backported QSPI implementation no longer works with recent versions of espressif core.

The default strategy is to wait for espressif to implement [QSPI support with arduino-esp32 3.1.0](https://github.com/espressif/arduino-esp32/issues/7063#issuecomment-1954558942), however espressif missed that milestone. That strategy requires to find a workaround for the missing initQuad() in Arduino SPI library.

The alternate strategy is to follow @moononournation's [Arduino_GFX](https://github.com/moononournation/Arduino_GFX/blob/master/src/databus/Arduino_ESP32QSPI.cpp) implementation, where QSPI uses esp-idf functions instead of relying on Arduino SPI library. However, that strategy requires that any **non-panel** bus driver also bypasses the Arduino SPI library to prevent any bus conflict. Consequently, going into that direction creates a significant tech debt and adds some limitations.

### QSPI Panel devices
- SH8601Z
- SH8501
- RM67162
- RM690B0
- NV3041A

### MIPI Panel devices
- AXS15231B Controller

### Known products
- LilyGo Amoled series (T-Display-S3, T4-S3, etc)
- CYD (JC3248W535EN, JC4827W543_4.3inch_ESP32S3)
- M5Stack Tab5 (ILI9881C MIPI-DSI)

### Pull Requests
- https://github.com/lovyan03/LovyanGFX/pull/388

### Issues
- https://github.com/lovyan03/LovyanGFX/issues/463
- https://github.com/lovyan03/LovyanGFX/issues/467
- https://github.com/lovyan03/LovyanGFX/issues/547
- https://github.com/lovyan03/LovyanGFX/issues/549
- https://github.com/lovyan03/LovyanGFX/issues/662
- https://github.com/lovyan03/LovyanGFX/issues/609

### Discussions
- https://github.com/lovyan03/LovyanGFX/discussions/604
- https://github.com/lovyan03/LovyanGFX/discussions/660

### Persons of interest
- @jgauchia
- @Westcott1
- @Starbane
- @FeralAI
- @nikthefix
- @zeroomega
- @powersoft
- @Forairaaaaa
- @Shadowtrance
- @moononournation
- @lovyan03
- @tobozo

**Comment by jgauchia** [comment]
Thank you @tobozo ....
I'm trying to understand how Arduino GFX init QSPI bus, and try to follow QSPI branch of lovyan and init RM690B0 panel.... 
Sorry, I don't have the same programming level that yours (or other members)... but I'm available for testing, o another things...
Likewise, I will try to do some tests and try "to do" something.

**Comment by Shadowtrance** [comment]
Would love proper support for QSPI displays as they seem to be getting more common. And i really like this library. :)
I wish comments were in english but that's nothing but a quick google translate doesn't fix when needed and I'm not going to hold it against someone when their native language isn't english. :)

And it would be very useful because a lot of projects out there use TFT_eSPI (which doesn't support it) or LFGX (which i prefer) as it's basically the same but better.

For my particular display, the JC3248W535EN with the AXS15231B controller as i mentioned in my request for help works with Arduino_GFX and [esp-idf](https://components.espressif.com/components/espressif/esp_lcd_axs15231b)

Some more info on the screen [here](https://github.com/NorthernMan54/JC3248W535EN/tree/main/JC3248W535EN) and the [controller datasheet](https://dl.espressif.com/AE/esp_iot_solution/AXS15231B_Datasheet_V0.5_20230306.pdf) and [some more here](https://github.com/espressif/esp-iot-solution/tree/master/components/display/lcd/esp_lcd_axs15231b).

So yeah I'm willing to test whatever i can throw at my screen, my coding skills aren't great but i can get by. I'm more of a tinkerer that likes making things and sometimes destroying things. 🤣

**Reply by tobozo** [reply]
That AXS15231B datasheet is both fascinating and scary :sweat_smile: 

> I'm more of a tinkerer that likes making things and sometimes destroying things. 🤣

"if it ain't broken, fix it until it is"

**Reply by Shadowtrance** [reply]
Came across this while seeing if anything new/interesting to do with my screen had popped up. Probably not really helpful in any way but thought it was cool even though I don't really like python lol [micropython_lcd](https://github.com/straga/micropython_lcd/tree/master/device/JC3248W535/driver/axs15231b)

**Comment by nikthefix** [comment]
@tobozo I also find it interesting that so many of the once MIPI only amoled displays are now supporting QSPI. Could it be a nod to the ever growing maker community? Companies like DWO now not only offer QSPI variants but they also integrate the power driver circuitry so we don't have to.

nik

**Comment by nikthefix** [comment]
@Shadowtrance Lovyan does offer proper support for QSPI displays - just maybe not the specific model the developer chooses. One thing I really like about Lovyan is that it exposes the hardware config (pin map, timings etc) at the sketch / main() level. You only need to delve into the library code to modify the init code sequence for a particular display (minimal modification at worst in my experience for amoled displays). And of course you'd need to make this modification to any GFX library which didn't have a specific template for your chosen display. For those using LVGL then all GFX lib options are overkill as the drawing functions are redundant since only the 'write pixels from frame buffer' function is utilized. In this case it makes more sense to distill the bus / pin / qspi setup functions from your preferred lib and roll your own driver. I have examples of this - largely drawn from Lovyan - on my github (which needs updating BTW)

nik

**Reply by Shadowtrance** [reply]
I guess "proper" was the wrong word (either way, i didn't mean it in a bad way). But i guess i meant just more support for these displays in general. And i dug into the code a lot and other code too and the result is only partially functional on my display. See my discussion asking for help. :)
But yeah, I'm all for tinkering and trying to figure things out. :)

**Reply by nikthefix** [reply]
No this is all productive talk.
Maybe I can help. I recently transferred some code for a Lilygo board using Arduino_GFX to Lovyan. 
The text above doesn't really provide enough information. 
Could you provide more? board, goal, bus type (qspi), display type, IDE?

nik

**Reply by Shadowtrance** [reply]
https://github.com/lovyan03/LovyanGFX/discussions/660
Board/Display is JC3248W535EN with the AXS15231B controller.

Goal... to be able to use LGFX (and sorta TFT_eSPI) and all the cool projects out there that use them, bus type QSPI, display type ?, IDE platformio arduino.

One project I'd like to get running is FreeTouchDeck...but it uses TFT_eSPI for some parts but there are LGFX forks floating around.
But yeah it would be nice to be able to use it without getting a headache. haha

It's a nice little screen but the display/touch controller is a pain in the butt...atm anway. I don't know if it's a reasonably new chip or not but I'd never heard of it before buying this screen.

**Reply by tonbor** [reply]
I have the JC3248W535EN with the AXS15231B controller, did you succeed in get it going with lovyanGFX library, and so do you have the right panel files?

**Comment by tobozo** [comment]
QSPI support has [landed on the develop branch](https://github.com/lovyan03/LovyanGFX/commit/7c8fd6205501a992bcb3394916d51bc3aed30adb), it is now part of **Bus_SPI** and is enabled via configuration:

The QSPI branch will no longer be updated, please switch to the develop branch for any further feedback.

:warning: Existing custom configurations should rename **Bus_QSPI** to **Bus_SPI**.

example with Lilygo T-Display-S3 AMOLED: 

```c

class Lilygo_T_Display_S3_AMOLED : public lgfx::LGFX_Device
{

  lgfx::Bus_SPI       _bus_instance;
  lgfx::Panel_RM67162 _panel_instance; // 1.91 inch RM67162 IPS AMOLED:

  public:

    Lilygo_T_Display_S3_AMOLED(void)
    {
      {
        auto cfg = _bus_instance.config();

        cfg.freq_write = 75000000;
        cfg.freq_read  = 20000000; 

        cfg.pin_sclk  = GPIO_NUM_47;

        cfg.pin_io0   = GPIO_NUM_18; // <<< quad spi pin
        cfg.pin_io1   = GPIO_NUM_7; // <<< quad spi pin
        cfg.pin_io2   = GPIO_NUM_48; // <<< quad spi pin
        cfg.pin_io3   = GPIO_NUM_5; // <<< quad spi pin

        cfg.spi_host   = SPI2_HOST;
        cfg.spi_mode   = SPI_MODE0;
        cfg.dma_channel = SPI_DMA_CH_AUTO;

        _bus_instance.config(cfg);
        _panel_instance.setBus(&_bus_instance);
      }

      // ....
  }

};
```

**Comment by jgauchia** [comment]
I'm testing develop branch with t4-s3 now.... Is it known when it will be included in the master branch or in a future release?

Thank you

**Comment by tobozo** [comment]
MIPI-DSI experimental support for ESP32P4 landed on [develop branch](https://github.com/lovyan03/LovyanGFX/tree/develop), there's a config block available for M5Stack Tab5.

![image](https://github.com/user-attachments/assets/9d8f9ebf-480f-4b64-b8c2-e733c440af1e)

**Reply by brightproject** [reply]
> MIPI-DSI experimental support for ESP32P4 landed on [M5GFX_Backport branch](https://github.com/lovyan03/LovyanGFX/tree/M5GFX_Backport), there's a config block available for M5Stack Tab5.

The branch has disappeared or was stolen by aliens😁

Hello @tobozo 
I have a device with a display `JC3636W518` connected via a bus `QSPI`.
I successfully connect the `LVGL` library to the device using the panel from `ESP-IDF`.

> #include <ESP_IOExpander.h
> #include <ESP_Panel_Library.h>

I also successfully connect the `Arduino_GFX` library

> https://github.com/moononournation/Arduino_GFX/blob/master/src/display/Arduino_ST77916.h

As you can see in the photo, the display works well.

![389305721-c30fda86-2164-4bd6-963d-d5439ef70772](https://github.com/user-attachments/assets/7682de4b-1c81-4ca2-be44-90b68de3297c)

I want use the `LovyanGFX` library and successfully connected the display via `RGB` _Panel_ST7701_guition_esp32_4848S040_.

![5433934873225195396](https://github.com/user-attachments/assets/5cb55dc6-e340-4d86-bd0e-fd2ea4ea1c03)

But I had difficulties connecting the display to `LovyanGFX` via `QSPI` - _ST77916_.
I am using version `1.2.7`, to which I added some code from the `QSPI` branch from the files related to the `LGFX_USE_QSPI` flag

> https://github.com/lovyan03/LovyanGFX/blob/QSPI/src/lgfx/v1/platforms/esp32/Bus_SPI.cpp
> http://github.com/lovyan03/LovyanGFX/blob/QSPI/src/lgfx/v1/platforms/esp32/common.cpp

I also found this file in the panel files

> lgfx/v1/panel/Panel_ST77961.hpp

At first I thought it was a different display than my `ST77916`, that is, only 2 digits `16` and `61` were changed.
I created my own file

> lgfx/v1/panel/Panel_ST77916.hpp

based on this file

> https://github.com/moononournation/Arduino_GFX/blob/master/src/display/Arduino_ST77916.h

The code is loaded into the device, the display backlight works, the code inside the device also works and I receive data in the serial port, only the device initialization does not occur.
I connect the panel like this:
```
#ifndef DISPLAY_QSPI_DRIVER_H
#define DISPLAY_QSPI_DRIVER_H

// #define LGFX_USE_QSPI

// #include "Arduino.h"
#include <LovyanGFX.h>
// #include <lgfx/v1/panel/Panel_ST77916.hpp>
#include <lgfx/v1/panel/Panel_ST77961.hpp>

#if !defined LGFX_USE_QSPI
  #error "This device does not support QSPI"
#endif

#define TFT_SPI_FREQ_HZ (30 * 1000 * 1000) // QSPI 30MHz
#define FREQ 500         // PWM frequency

class LGFX : public lgfx::LGFX_Device {

  lgfx::Bus_SPI       _bus_instance;
  // lgfx::Panel_ST77916 _panel_instance;
  lgfx::Panel_ST77961 _panel_instance;
  // lgfx::Bus_QSPI       _bus_instance;
  lgfx::Touch_CST816S   _touch_instance;
  lgfx::Light_PWM     _light_instance;

public:

    LGFX(void)
    {
        {
            auto cfg = _bus_instance.config();
            
            cfg.freq_write  = TFT_SPI_FREQ_HZ;
            cfg.freq_read   = 16000000;
            
            cfg.spi_host    = SPI2_HOST;
            cfg.spi_mode    = SPI_MODE0;
            // cfg.spi_3wire   = true;
            // cfg.use_lock    = true;
            cfg.dma_channel = SPI_DMA_CH_AUTO;

            cfg.pin_sclk = GPIO_NUM_9; // QSPI SCK pin
            // cfg.pin_dc = GPIO_NUM_10; // QSPI CS/DC pin
            cfg.pin_io0  = GPIO_NUM_11; // QSPI D0 pin
            cfg.pin_io1  = GPIO_NUM_12; // QSPI D1 pin
            cfg.pin_io2  = GPIO_NUM_13; // QSPI D2 pin
            cfg.pin_io3  = GPIO_NUM_14; // QSPI D3 pin

            _bus_instance.config(cfg);
            _panel_instance.setBus(&_bus_instance);
        }


        {
            auto cfg = _panel_instance.config();

            cfg.pin_cs   = GPIO_NUM_10; // Chip select pin
            cfg.pin_rst  = GPIO_NUM_47 ; // Reset pin
            cfg.pin_busy = -1;          // Busy pin (not used)

            cfg.panel_width   = 360;
            cfg.panel_height  = 360;

            cfg.memory_width  = 360;
            cfg.memory_height = 360;

            cfg.offset_x      = 0;
            cfg.offset_y      = 0;

            cfg.offset_rotation = 0;

            cfg.dummy_read_pixel = 8;
            cfg.dummy_read_bits  = 1;

            cfg.readable   = true;
            cfg.invert     = true;
            cfg.rgb_order  = true;
            cfg.dlen_16bit = false;
            cfg.bus_shared = true;

            _panel_instance.config(cfg);
        }

        {  
            auto cfg = _touch_instance.config();

            cfg.i2c_port = 0;
            cfg.i2c_addr = 0x15; // I2C address of the CST816S controller
            cfg.pin_sda  = GPIO_NUM_7; // 7
            cfg.pin_scl  = GPIO_NUM_8; // 8
            cfg.pin_int  = GPIO_NUM_41; // 41
            // cfg.pin_int  = GPIO_NUM_NC; // -1
            cfg.pin_rst  = GPIO_NUM_40; // 40
            cfg.freq     = 400000;
            
            cfg.x_min = 0;
            cfg.y_min = 0;
            cfg.x_max = 360 - 1;
            cfg.y_max = 360 - 1;

            _touch_instance.config(cfg);
            _panel_instance.setTouch(&_touch_instance);
        }

        {
            auto cfg = _light_instance.config();

            cfg.pin_bl = GPIO_NUM_15 ;  // Backlight pin 15
            cfg.freq = FREQ;
            cfg.pwm_channel = LEDC_CHANNEL_0;
            // cfg.invert = false;

            _light_instance.config(cfg);
            _panel_instance.setLight(&_light_instance);
        }

        setPanel(&_panel_instance);
    }

};

#endif // DISPLAY_QSPI_DRIVER_H
```

I changed the panel to `Panel_ST77961`, but there is no image on the display.

Based on this information

> https://github.com/lovyan03/LovyanGFX/pull/612

Downloaded the developer branch, where version `1.2.9`, but there is no result, the display does not show anything.

I also read this thread

> https://github.com/lovyan03/LovyanGFX/discussions/660

I understand that the problem with using `QSPI` in the library `LovyanGFX` is quite old and as I understand it does not have a solution for all devices.

I would like to understand what I am doing wrong and find a solution to the problem.

**Comment by tobozo** [comment]
@brightproject both QSPI and M5GFX backports are now available on the develop branch, QSPI branch will soon be deleted too.

I suggest you extend Panel_AMOLED instead of Panel_LCD, see Panel_RM690B0.hpp as an example.

**Reply by brightproject** [reply]
> I suggest you extend Panel_AMOLED instead of Panel_LCD, see Panel_RM690B0.hpp as an example.

Unfortunately, when I specify in the code

> struct Panel_ST77916 : public Panel_AMOLED

The image is still missing, but the display backlight, which worked with the code

> struct Panel_ST77916 : public Panel_LCD

also disappears

**Reply by brightproject** [reply]
I don't quite understand the logic of the library `LovyanGFX`.

Why is sometimes the panel configuration one header file, like for example:

> https://github.com/lovyan03/LovyanGFX/blob/develop/src/lgfx/v1/panel/Panel_ST77961.hpp

And sometimes it's two files, both the `*h` and the `*cpp`, like for example:

> https://github.com/lovyan03/LovyanGFX/blob/develop/src/lgfx/v1/panel/Panel_SSD1963.hpp
> https://github.com/lovyan03/LovyanGFX/blob/develop/src/lgfx/v1/panel/Panel_SSD1963.cpp

As I already mentioned, I don't really understand how `LovyanGFX` works in terms of working with panels, but otherwise I see similarities with `TFT_eSPI`.

But I still understand that the display is not initialized now because I don't specify it.

@tobozo Could you help me please with `ST77916` display driver by `QSPI` for `LovyanGFX,` based on these files🙂?

> https://github.com/esp-arduino-libs/ESP32_Display_Panel/blob/master/src/drivers/lcd/port/esp_lcd_st77916.h
> https://github.com/esp-arduino-libs/ESP32_Display_Panel/blob/master/src/drivers/lcd/port/esp_lcd_st77916.c

**Comment by brightproject** [comment]
Hello @tobozo 🙂
Solved some problems thanks to @jgauchia 

> https://github.com/lovyan03/LovyanGFX/issues/547#issuecomment-3052618375

There is some news that `LovyanGFX` will work on?

> framework-arduinoespressif32 @ 3.2.1

**Comment by brightproject** [comment]
Sorry @tobozo , why does calling `Light_PWM` not work?
```
lgfx::Light_PWM     _light_instance;
{
    auto cfg = _light_instance.config();

    cfg.pin_bl = GPIO_NUM_15;  // Backlight pin 15
    // cfg.pin_bl = GPIO_NUM_NC;  // NO backlight pin -1
    cfg.freq = FREQ;
    cfg.pwm_channel = LEDC_CHANNEL_0;
    // cfg.pwm_channel = 0;
    // cfg.invert = false;

    _light_instance.config(cfg);
    _panel_instance.setLight(&_light_instance);
}
```
The backlight can only be started by pin `BL` in high.
```
lgfx::pinMode(GPIO_NUM_15, lgfx::pin_mode_t::output); // power enable for LCD
lgfx::Light_PWM;
lgfx::gpio_hi(GPIO_NUM_15);
```

This is both inconvenient and incomprehensible at the same time, and such a strange work only for `Panel_AMOLED`🙄

**Comment by brightproject** [comment]
I made the backlight myself using `PWM`

```c++
#include "Arduino.h"

// When building under framework-arduinoespressif32 2.0.17, you need to disable the flag
#define CORE_3_2_1

// PWM settings
#define LED_PIN GPIO_NUM_15
#define LED_CHANNEL 7 // PWM channel
#define FREQ 1000 // PWM frequency (1 kHz)
#define RESOLUTION 8 // Resolution (8 bits)
// #define RESOLUTION 12 // Resolution (12 bits)

void setup() {
  #ifdef CORE_3_2_1
    ledcAttachChannel(LED_PIN, FREQ, RESOLUTION, LED_CHANNEL); // 0-255
    // ledcAttachChannel(LED_PIN, FREQ, RESOLUTION, LED_CHANNEL); // 0-4095
    ledcWrite(LED_PIN, 255); // 0-255
    // ledcWrite(LED_PIN, 4095); // 0-4095
  #else
    ledcSetup(LED_CHANNEL, FREQ, RESOLUTION);
    ledcAttachPin(LED_PIN, LED_CHANNEL);
    ledcWrite(LED_PIN, 255);
  #endif
} // SETUP

void loop() {

} // LOOP
```
